decred/decred/util/helpers.py

```diff
@@ -236,14 +236,43 @@ def formatNumber(number, billions="B", spacer=" ", isMoney=False):
     return ("%.2e%s" % (flt, spacer)).replace("e-0", "e-")
 
 
-rootLogger = logging.getLogger("")
-rootLogger.setLevel(logging.NOTSET)
+class LogSettings:
+    """Used to track a few logging-related settings"""
 
+    root = logging.getLogger("")
+    defaultLevel = logging.INFO
+    moduleLevels = {}
+    loggers = {}
 
-def prepareLogger(name, filepath=None, logLvl=logging.INFO):
-    """
-    Set logger settings appropriately
+
+LogSettings.root.setLevel(logging.NOTSET)
+
+
+def prepareLogging(filepath=None, logLvl=logging.INFO, lvlMap=None):
     """
+    Prepare for using getLogger. Logs to stdout. If filepath is provided, log
+    outputs will be saved to a rotating log file at the specified location. Any
+    loggers, both future loggers and those already created, will have their
+    levels set according to the new logLvl and lvlMap.
+
+    Args:
+        filepath (str or pathlib.Path): optional. The base name for the rotating
+            log file.
+        logLvl (int): optional. default logging.INFO. The default logging level
+            used for all new loggers without entries in the lvlMap.
+        lvlMap: (dict): optional. If provided, the name->level mapping will be
+            added to the stored level dict, which is referenced when loggers are
+            created using getLogger.
+    """
+    # Set log level for existing loggers.
+    LogSettings.defaultLevel = logLvl
+    recursiveUpdate(LogSettings.moduleLevels, lvlMap if lvlMap else {})
```

Comment:
This seems to be the first use in the codebase of `recursiveUpdate` (it was already there, unused). Its docstring says:

> Recursively update the target dictionary with the source dictionary, leaving unfound keys in place. This is different than `dict.update`, which removes target keys not in the source.

This is not the case, as far as I can see. I could not come up with an example where `dict.update` removed target keys not in the source. (Such behavior is also not described in the docs, and would be very surprising).

`dict.update` seems to handle recursive dicts correctly too, so I currently don't see a reason for the existence and use of `recursiveUpdate`.

Reply:
I think you're right that it's not needed and `dict.update` works for our needs. I don't know what I was thinking when I wrote that, but the purpose of `recursiveUpdate` was to update dicts with dict values recursively. 

```
d = {
	"a": 1,
	"b": {
		"c": 3,
		"d": 4,
	}
}
```
with `update`

```
d.update({"b": {"e": 5}})

{
	"a": 1,
	"b": {
		"e": 5,
	}
}
```

but with `recursiveUpdate`

```
recursiveUpdate(d, {"b": {"e": 5}})

d = {
	"a": 1,
	"b": {
		"c": 3,
		"d": 4,
		"e": 5,
	}
}
```

Buy like you said, we don't appear to be using it and it's not needed in this case anyway. Will remove.

Reply:
Ah, there you go! In my attempts I did add multiple levels to the source, but did not do so with the target. It still seems surprising behavior for a stdlib type: I'd bet there's been discussion about this before.